src/bin/cargo/cli.rs

```diff
@@ -33,9 +36,17 @@ pub fn main(config: &mut Config) -> CliResult {
     let args = match cli().try_get_matches() {
         Ok(args) => args,
         Err(e) => {
-            if e.kind == clap::ErrorKind::UnrecognizedSubcommand {
+            if e.kind() == clap::ErrorKind::UnrecognizedSubcommand {
                 // An unrecognized subcommand might be an external subcommand.
-                let cmd = e.info[0].clone();
+                let cmd = e
+                    .context()
+                    .find_map(|c| match c {
+                        (ContextKind::InvalidSubcommand, &ContextValue::String(ref cmd)) => {
+                            Some(cmd)
+                        }
+                        _ => None,
+                    })
```

Comment:
Clap feature request: this would be way better as:

```rust
let cmd = e.context::<clap::context::InvalidSubcommand>().expect(...);
```

where the signature is:

```rust
impl Error {
    fn context<Kind: ContextKind>(&self) -> Option<&Kind::ContextValue>;
}
```

It seems clumsy that to access the piece of context you want requires iterating arbitrary other context kinds and having to assert that clap didn't stick your context in there with a type that it wasn't supposed to.

Reply:
Yeah, this code made me really sad to write.

The proposed interface may have to be able to deal with multiple values for a given kind, I'm not sure, but worth thinking about for sure 👍

Reply:
I started with the most general API, awaiting seeing how it was going to be used to know what would be helpful.

Since `fn context()` is already in use, we'll just have to come up with a different name for this.